src/linear_algebra/bilinear_form.lean

```diff
@@ -400,3 +408,194 @@ begin
 end
 
 end alt_bilin_form
+
+namespace bilin_form
+
+section endomorphism_adjoints
+
+variables {R : Type u} [comm_ring R]
+variables {M : Type v} [add_comm_group M] [module R M]
+variables (B : bilin_form R M)
+
+/-- Given a module equipped with a bilinear form, this is the condition for a pair of endomorphims
+to be mutually adjoint. -/
+def is_adjoint_pair (T S : module.End R M) := ∀ x y, B (T x) y = B x (S y)
```

Comment:
Hmm, sorry that I didn't notice this in an earlier review. But shouldn't adjoint maps be generalised to `f : M → N` and `g : N → M` with bilinear forms on `M` and `N`?

Reply:
OK sure. I had considered this. I’ll adjust and add some definitions for perfect pairings etc this evening!

Reply:
Thinking about this a little more, I remember why I shied away from generalising to a pair of modules: I'm a bit unsure about singling out one particular generalisation in this situation.

Of course the generalisation you highlight is natural and useful. The data is an element of each of:
 - `Hom(M ⊗ M, R)`
 - `Hom(N ⊗ N, R)`
 - `f ∈ Hom(M, N)`
 - `g ∈ Hom(N, M)`
Given this, we have two natural ways to build elements of `Hom(M ⊗ N, R)` and we require that they are equal. But since the forms are not required to be symmetric, we could instead require that the two natural ways to build elements of `Hom(N ⊗ M, R)` are equal. So actually we have a left- and right- handed generalisations. Of course swapping `M` and `N` swaps left and right so we could just single out one way. Also, this complication exists with only a single module (carrying bilinear form) but it feels like it's less of an issue since the bilinear form sits in the background and we just swap `f` and `g`.

More seriously, there are generalisations with different data. For example, we could take our data to be an element of each of:
 - `Hom(M ⊗ N, R)`
 - `Hom(M, M)`
 - `Hom(N, N)`
Given this, we have two ways to build new elements of `Hom(M ⊗ N, R)` and we can require that they are equal.

So it feels to me that by allowing a pair of modules, there is a sudden explosion from once concept to a whole family and it is not totally obvious what to define.

What do you think? Am I bikeshedding or missing something obvious? Should I just define the left-handed property associated with the data you suggest and get on with it? It feels like the happy medium might be to restrict to either symmetric forms or to a single module. Do you have a preference between these two options?

Reply:
Hmm, I think this merits a discussion on Zulip. Thanks for looking into this. If other people don't chime in, I will try to help beginning of next week.

Reply:
https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.232746.20adjoints.20of.20linear.20maps

Reply:
On reflection, I think I was being a bit ridiculous above. I have just pushed a commit which introduces your original suggestion.

However it has given me another idea: I may be able to neatly eliminate a lot of the self / skew repetition by proving results for just the self adjoint case but with a pair of bilinear forms on the underlying module. When the pair of forms are just the same form repeated, then this is the true self adjoint case and when the pair of forms are a form and its negation then this is the true skew adjoint case (see `is_skew_adjoint_iff_neg_self_adjoint` in most recent commit).

So I think I'll mark this PR as draft until I have experimented with this tomorrow evening.

Thanks again for this suggestion.

Reply:
Now done.